libgeopmd/src/PowercapSysfsDriver.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2015 - 2025 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "PowercapSysfsDriver.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cmath>
+#include <cstring>
+#include <sstream>
+
+#include "geopm/Helper.hpp"
+
+#include "SysfsIOGroup.hpp"
+#include "geopm/PlatformTopo.hpp"
+
+static const std::string POWERCAP_DIRECTORY = "/sys/class/powercap";
+
+static std::map<std::string, std::string> load_powercap_resource_by_name(const std::string &powercap_directory)
+{
+    std::map<std::string, std::string> result;
+
+    for (const auto &policy_file : geopm::list_directory_files(powercap_directory)) {
+        if (policy_file.find("intel-rapl:") != 0) {
+            continue;
+        }
+
+        std::ostringstream oss;
+        oss << powercap_directory << "/" << policy_file << "/name";
+        auto name_path = oss.str();
+
+        int name_fd = open(name_path.c_str(), O_RDONLY);
+        if (name_fd == -1) {
+            throw geopm::Exception("PowercapSysfsDriver failed to open " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        char name_buf[geopm::SysfsDriver::M_IO_BUFFER_SIZE] = {};
+        int read_bytes = pread(name_fd, name_buf, sizeof(name_buf) - 1, 0);
+        close(name_fd);
+        if (read_bytes < 0) {
+            throw geopm::Exception("PowercapSysfsDriver failed to read " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        if (read_bytes == sizeof(name_buf) - 1) {
+            throw geopm::Exception("PowercapSysfsDriver truncated read from " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+
+        std::string name(name_buf);
+        if (!name.empty() && name.back() == '\n') {
+            name.pop_back();
+        }
+        result.emplace(name, policy_file);
+    }
+
+    return result;
+}
+
+namespace geopm
+{
+    const std::string powercap_sysfs_json(void);
+
+    PowercapSysfsDriver::PowercapSysfsDriver()
+        : PowercapSysfsDriver(POWERCAP_DIRECTORY)
+    {
+    }
+
+    PowercapSysfsDriver::PowercapSysfsDriver(const std::string &powercap_directory)
+        : M_PROPERTIES{SysfsDriver::parse_properties_json(plugin_name(), powercap_sysfs_json())}
+        , M_POWERCAP_RESOURCE_BY_NAME(load_powercap_resource_by_name(powercap_directory))
+        , m_powercap_directory(powercap_directory)
+    {
+    }
+
+    int PowercapSysfsDriver::domain_type(const std::string &name) const
+    {
+        if (M_PROPERTIES.find(name) == M_PROPERTIES.end()) {
+            throw Exception("PowercapSysfsDriver: unknown name: " + name,
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return GEOPM_DOMAIN_PACKAGE;
+    }
+
+    static std::string name_to_resource_key(const std::string &name, int domain_idx)
```

Comment:
Minor recommendation: could these static functions be moved to the top (to facilitate reading the code)? It might just throw someone off a bit when they're going through class functions, and then encounter a static function not part of the class. It might just be me, so feel free to ignore this suggestion.